power_grid_model_c/power_grid_model/include/power_grid_model/main_model.hpp

```diff
@@ -732,218 +647,25 @@ class MainModelImpl<ExtraRetrievableTypes<ExtraRetrievableType...>, ComponentLis
-                        return power_sensor.get_output<sym>(math_output[obj_math_id.group].source[obj_math_id.pos].s);
-                    case shunt:
-                        return power_sensor.get_output<sym>(math_output[obj_math_id.group].shunt[obj_math_id.pos].s);
-                    case load:
-                    case generator:
-                        return power_sensor.get_output<sym>(math_output[obj_math_id.group].load_gen[obj_math_id.pos].s);
-                    case node:
-                        return power_sensor.get_output<sym>(
-                            math_output[obj_math_id.group].bus_injection[obj_math_id.pos]);
-                    default:
-                        throw MissingCaseForEnumError(std::string(GenericPowerSensor::name) + " output_result()",
-                                                      terminal_type);
-                }
-            });
-    }
-
-    // output fault
-    template <bool sym, class Component, class ResIt>
-    std::enable_if_t<
-        std::is_base_of_v<std::forward_iterator_tag, typename std::iterator_traits<ResIt>::iterator_category> &&
-            std::is_same_v<Fault, Component>,
-        ResIt>
-    output_result(std::vector<MathOutput<sym>> const& /* math_output */, ResIt res_it) {
-        assert(construction_complete_);
-        return std::transform(components_.template citer<Component>().begin(),
-                              components_.template citer<Component>().end(), comp_coup_->fault.cbegin(), res_it,
-                              [](Fault const& fault, Idx2D /* math_id */) {
-                                  return fault.get_output();
-                              });
-    }
-
-    template <bool sym>
-    void output_result(std::vector<MathOutput<sym>> const& math_output, Dataset const& result_data, Idx pos = 0) {
-        static constexpr std::array<OutputFunc<sym>, n_types> get_result{
-            [](MainModelImpl& model, std::vector<MathOutput<sym>> const& math_output_,
+    template <math_output_type MathOutputType>
+    void output_result(std::vector<MathOutputType> const& math_output, Dataset const& result_data, Idx pos = 0) {
+        static constexpr std::array<OutputFunc<MathOutputType>, n_types> get_result{
+            [](MainModelImpl& model, std::vector<MathOutputType> const& math_output_,
                DataPointer<false> const& data_ptr, Idx position) {
                 auto const begin =
-                    data_ptr.get_iterators<typename ComponentType::template OutputType<sym>>(position).first;
-                model.output_result<sym, ComponentType>(math_output_, begin);
+                    data_ptr
+                        .get_iterators<std::conditional_t<
+                            steady_state_math_output_type<MathOutputType>,
+                            typename ComponentType::template OutputType<symmetric_math_output_type<MathOutputType>>,
+                            typename ComponentType::ShortCircuitOutputType>>(position)
```

Comment:
i know this looks really ugly but the alternative is to also change all `OutputType` and `ShortCircuitOutputType`s in the individual components to support the calculation type logic and i would like to do that in a separate PR to prevent too many things from being touched at once